airbyte-cdk/python/unit_tests/sources/test_source.py

```diff
@@ -93,18 +104,185 @@ def streams(self, config):
+                            ],
+                        ),
+                    }
+                ),
+            ],
+            None,
+        ),
+        (
+            "test_incoming_legacy_state",
+            {"movies": {"created_at": "2009-07-19"}, "directors": {"id": "villeneuve_denis"}},
+            [
+                AirbyteStateMessage(
+                    type=AirbyteStateType.LEGACY, data={"movies": {"created_at": "2009-07-19"}, "directors": {"id": "villeneuve_denis"}}
+                )
+            ],
+            None,
+        ),
+        ("test_empty_incoming_stream_state", [], [], None),
+        ("test_none_incoming_state", None, [], None),
+        ("test_empty_incoming_legacy_state", {}, [], None),
+        (
+            "test_invalid_stream_state_invalid_type",
+            [
+                {
+                    "type": "NOT_REAL",
+                    "stream": {
+                        "stream_state": {"created_at": "2009-07-19"},
+                        "stream_descriptor": {"name": "movies", "namespace": "public"},
+                    },
+                }
+            ],
+            None,
+            ValidationError,
+        ),
+        (
+            "test_invalid_stream_state_missing_descriptor",
+            [{"type": "STREAM", "stream": {"stream_state": {"created_at": "2009-07-19"}}}],
+            None,
+            ValidationError,
+        ),
+        (
+            "test_invalid_global_state_missing_streams",
+            [{"type": "GLOBAL", "global": {"shared_state": {"shared_key": "shared_val"}}}],
+            None,
+            ValidationError,
+        ),
+        (
+            "test_invalid_global_state_streams_not_list",
+            [
+                {
+                    "type": "GLOBAL",
+                    "global": {
+                        "shared_state": {"shared_key": "shared_val"},
+                        "stream_states": {
+                            "stream_state": {"created_at": "2009-07-19"},
+                            "stream_descriptor": {"name": "movies", "namespace": "public"},
+                        },
+                    },
+                }
+            ],
+            None,
+            ValidationError,
+        ),
+        ("test_invalid_state_message_has_no_stream_global_or_data", [{"type": "LEGACY", "not": "something"}], None, ValidationError),
+    ],
+)
+def test_read_state(source, test_name, incoming_state, expected_state, expected_error):
     with tempfile.NamedTemporaryFile("w") as state_file:
-        state_file.write(json.dumps(state))
+        state_file.write(json.dumps(incoming_state))
         state_file.flush()
-        actual = source.read_state(state_file.name)
-        assert state == actual
+        if expected_error:
+            with pytest.raises(expected_error):
+                source.read_state(state_file.name)
+        else:
+            actual = source.read_state(state_file.name)
+            assert actual == expected_state
```

Comment:
nit: What do you think about using `does_not_raise` for a cleaner conditional raising?
```suggestion
       # If expected_error == does_not_raise pytest will not expect an error to be raised and will continue the test execution
        with pytest.raises(expected_error):
            source.read_state(state_file.name)
            actual = source.read_state(state_file.name)
            assert actual == expected_state
```

Reply:
I'm not entirely sure I understand the code suggestion you posted. The way it's written, `pytest.raises(expected_error)` expects an iterable type, so the test will fail on any happy path case since the expected_error is None. Wouldn't these tests also fail since w/o the condition, we'd always be expecting some type of error which isn't the case for valid state input?

Reply:
actually reading more about this do you mean, that for all happy path cases as test parameters we use `does_not_raise` and then when checking for errors `pytest.raises(Exception`. And code looking something like this:
```
        with expected_error:
            actual = source.read_state(state_file.name)
            assert actual == expected_state
```

Reply:
Exactly! (example [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/source-acceptance-test/unit_tests/test_core.py#L150))